feat: migration guide (#163)
matter-labs/zksync-docs: content/00.build/40.tooling/20.hardhat/20.migrating-to-zksync.md

    ```bash [pnpm]
    pnpm i -D @matterlabs/hardhat-zksync
    ```

    ```bash [bun]
    bun add @matterlabs/hardhat-zksync
    ```

    ::

1. Import the plugin at the top of the `hardhat.config.ts` file:

    ```ts
    import "@matterlabs/hardhat-zksync";
    ```

1. Add the preferred ZKsync networks to the `hardhat.config.ts` file:

    ```js
    networks: {
        zkSyncSepoliaTestnet: {
          url: "https://sepolia.era.zksync.dev",
          ethNetwork: "sepolia",
          zksync: true,
          verifyURL: "https://explorer.sepolia.era.zksync.dev/contract_verification",
        },
        zkSyncMainnet: {
          url: "https://mainnet.era.zksync.io",
          ethNetwork: "mainnet",
          zksync: true,
          verifyURL: "https://zksync2-mainnet-explorer.zksync.io/contract_verification",
        },
        dockerizedNode: {
          url: "http://localhost:3050",
          ethNetwork: "http://localhost:8545",
          zksync: true,
        },
        inMemoryNode: {
          url: "http://127.0.0.1:8011",
          ethNetwork: "localhost",
          zksync: true,
        },
        // Other networks

    }
    ```

::callout{icon="i-heroicons-light-bulb"}

> You can also add the `zksync:true` flag to the `hardhat` network.

Author: It's optional. Only if you want to run tests with era-test-node

::

### Deployment with `hardhat-ethers`

The `@matterlabs/hardhat-zksync` includes `@matterlabs/hardhat-zksync-ethers`, a package that extends `@nomiclabs/hardhat-ethers`
with all the necessary helper methods to deploy contracts on both ZKsync and EVM networks. The injected `hre.ethers`
object provides methods like `deployContract`, `getContractFactory` or `getContractAt` so deployment scripts work out of the box.

::callout{icon="i-heroicons-exclamation-triangle" color="amber"}
To avoid typescript collision errors between `@nomiclabs/hardhat-ethers` and `@matterlabs/hardhat-zksync-ethers` make sure that only the latter
(or the wrapper plugin `@matterlabs/hardhat-zksync`) is imported in the `hardhat.config.ts` file.
::

See below examples for `hardhat-ethers` and `hardhat-zksync-ethers`:

::code-group

```typescript [hardhat-ethers]
const greeter = await hre.ethers.deployContract('Greeter', ['Hi there!']);
await greeter.waitForDeployment();
```

```typescript [hardhat-zksync-ethers]
const greeter = await hre.ethers.deployContract('Greeter', ['Hi there!']);
await greeter.waitForDeployment();
```

::

> In-case a custom deployment is needed, use `ContractFactory`.

Reviewer: ```suggestion
When a custom deployment is needed, use `ContractFactory`.
```

::code-group

```typescript [hardhat-ethers]
const GreeterFactory = await hre.ethers.getContractFactory('Greeter');
const greeterContract = GreeterFactory.deploy(); // if any, pass constructor arguments in deploy arguments
await greeter.waitForDeployment();
```

```typescript [hardhat-zksync-ethers]
const GreeterFactory = await hre.ethers.getContractFactory('Greeter');
const greeterContract = GreeterFactory.deploy(); // if any, pass constructor arguments in deploy arguments

> await greeter.waitForDeployment();

Reviewer: All of these code examples have the same code between the hardhat-ethers and hardhat-zksync-ethers. Is that expected?

Author: yes. I want to highlight that it's the same in EVM and ZKsync

The injected `hre.upgrades`
object provides methods like `deployProxy` so deployment scripts work out of the box.

::callout{icon="i-heroicons-exclamation-triangle" color="amber"}
To avoid typescript collision errors between `@openzeppelin/hardhat-upgrades` and `@matterlabs/hardhat-zksync-upgradable` make sure that only the latter
(or the wrapper plugin `@matterlabs/hardhat-zksync`) is imported in the `hardhat.config.ts` file.
::

See examples below:

#### Deployment (Proxy)

::code-group

```typescript [@openzeppelin/contracts-upgradable]
const constructorArguments = [...];
const initializerFunctionName = 'initialize'
const boxFactory = await hre.ethers.getContractFactory("Box");
const box = await hre.upgrades.deployProxy(boxFactory, constructorArguments, {
   initializer: initializerFunctionName,
});
await box.waitForDeployment();
```

```typescript [hardhat-zksync-upgradeable]
const constructorArguments = [...];
const initializerFunctionName = 'initialize'
const boxFactory = await hre.ethers.getContractFactory("Box");
const box = await hre.upgrades.deployProxy(boxFactory, constructorArguments, {
  initializer: initializerFunctionName
});

> await box.waitForDeployment();

Reviewer: Are these supposed to be the exact same?

Author: with the new version of the plugins, yes

  initializer: initializerFunctionName
});
await box.waitForDeployment();
```

::

#### Upgrade (Beacon)

::code-group

```typescript [@openzeppelin/hardhat-upgrades]
const constructorArguments = [...];
const initializerFunctionName = 'initialize'
const upgradableProxyAddress = "UPGRADEABLE_PROXY_ADDRESS";
const boxV2ContractFactory = await hre.ethers.getContractFactory("BoxV2");
const boxV2 = await hre.upgrades.upgradeProxy(upgradableProxyAddress, boxV2ContractFactory);
await boxV2.waitForDeployment();
```

```typescript [hardhat-zksync-upgradeable]
const constructorArguments = [...];
const initializerFunctionName = 'initialize'
const upgradableProxyAddress = "UPGRADEABLE_PROXY_ADDRESS";
const boxV2ContractFactory = await hre.ethers.getContractFactory("BoxV2");
const boxV2 = await hre.upgrades.upgradeProxy(upgradableProxyAddress, boxV2ContractFactory);
await boxV2.waitForDeployment();
```

::

### Custom deployment scripts with `hardhat-zksync`

Additionally, you can write custom deployment scripts for ZKsync leveraging the `hre.deployer` object which is injected automatically by `@matterlabs/hardhat-zksync-deploy`.

> The `Deployer` class provides helper methods to deploy smart contracts to ZKsync. Find [more info here](https://docs.zksync.io/build/tooling/hardhat/hardhat-zksync-deploy#methods).

Reviewer: ```suggestion
The `Deployer` class provides helper methods to deploy smart contracts to ZKsync. Learn more on [`hardhat-zksync-deploy` methods](/build/tooling/hardhat/hardhat-zksync-deploy#methods).
```

### Troubleshoting

- **Contract size too large:** if the size of the generated bytecode is too large and can not be deployed, try compiling
the contract with the `mode: 3` flag in the `hardhat.config.ts` file to

> optimize the bytecode size on compilation. [More info here](https://docs.zksync.io/build/tooling/hardhat/hardhat-zksync-solc#configuration).

Reviewer: ```suggestion
optimize the bytecode size on compilation. Learn more on [`hardhat-zksync-solc` configuration](/build/tooling/hardhat/hardhat-zksync-solc#configuration).
```

## Smart Contract Verification

<!-- TODO: pending rephrase -->

> For ZKsync, the verification of a smart contracts does not differ from the commonly used verification process.

Reviewer: ```suggestion
There are no differences in the verification of smart contracts on ZKsync.
```

The only thing required is to add a `verifyURL` on the ZKsync networks in the `hardhat.config.ts` file:

```typescript
  zkSyncSepoliaTestnet: {
    url: "https://sepolia.era.zksync.dev",
    ethNetwork: "sepolia",
    zksync: true,
    verifyURL: 'https://explorer.sepolia.era.zksync.dev/contract_verification'
  },
  zkSyncMainnet: {
    url: "https://mainnet.era.zksync.io",
    ethNetwork: "mainnet",
    zksync: true,
    verifyURL: "https://zksync2-mainnet-explorer.zksync.io/contract_verification",
  },
```

You can run the verification task programmatically inside a script as follows:

```js
const verificationId = await hre.run("verify:verify", {
  address: contractAddress,
  contract: contractFullyQualifedName,
  constructorArguments: [...]
});
```

Or execute the verification task from your terminal running:

```sh
npm hardhat verify --network testnet 0x7cf08341524AAF292255F3ecD435f8EE1a910AbF "Hi there!"
```

Find more info in [this section of the documentation](/build/tooling/hardhat/hardhat-zksync-verify).

## Scripting

<!-- TODO: Include mention to API compatibility in zksync-ethers and viem -->
To be included.

## Multichain projects

> Here are some recommendations projects that target multiple chains:

Reviewer: ```suggestion
Here are some recommendations for projects that target multiple chains:
```

- Add the desired ZKsync networks with the `zksync:true` flag to the `hardhat.config.ts`.
- Make sure to run the compilation task with the `--network` flag when targeting ZKsync networks to use the custom compiler.

> - The `@matterlabs/hardhat-zksync` plugin overrides the following plugins: `@nomiclabs/hardhat-ethers`, `@openzeppelin/hardhat-upgrades` when
> targeting ZKsync chains, while maintaining their original features when targeting EVM networks.

Reviewer: ```suggestion
- When targeting ZKsync chains, the `@matterlabs/hardhat-zksync` plugin overrides the following plugins: `@nomiclabs/hardhat-ethers`, `@openzeppelin/hardhat-upgrades`.
```

- To avoid typescript collision errors between `@nomiclabs/hardhat-ethers` and
`@openzeppelin/hardhat-upgrades` with  `@matterlabs/hardhat-zksync` make sure that only the latter is imported in the `hardhat.config.ts` file.

> - If you're using `hardhat-deploy` or `hardhat-ethers` your deployment scripts should work out of the box.

Reviewer: ```suggestion
- Your deployment scripts should not require any changes if you're using `hardhat-deploy` or `hardhat-ethers`.
```

> - If you have different deployment scripts for
> ZKsync, you can indicate the custom deployment folder for the ZKsync network using [the `deployPaths` property](/build/tooling/hardhat/hardhat-zksync-deploy#network-specific-deployment-paths):

Reviewer: ```suggestion
- If you have a separate directory for ZKsync deployment scripts, you can indicate the custom deployment folder for the ZKsync network using [the `deployPaths` property](/build/tooling/hardhat/hardhat-zksync-deploy#network-specific-deployment-paths):
```

```ts
const config: HardhatUserConfig = {

>   networks: {
>     sepolia: {
>       url: "https://sepolia.infura.io/v3/<API_KEY>",
>     },
>     zksyncTestnet: {
>       url: "https://sepolia.era.zksync.dev",
>       ethNetwork: "sepolia",
>       // ADDITION
>       deployPaths: "deploy-zksync", //single deployment directory
>       deployPaths: ["deploy", "deploy-zksync"], //multiple deployment directories
>       zksync: true,
>     },
>   },

Reviewer: ```suggestion
  networks: {
    zksyncTestnet: {
      url: "https://sepolia.era.zksync.dev",
      ethNetwork: "sepolia",
      zksync: true,
      // ADDITION
      deployPaths: "deploy-zksync", //single deployment directory
      deployPaths: ["deploy", "deploy-zksync"], //multiple deployment directories
    },
  },
```
